pkg/cmd/run.go

```diff
@@ -112,52 +124,114 @@ doppler run --command "YOUR_COMMAND && YOUR_OTHER_COMMAND"`,
 			}
 		}
 
-		secrets := fetchSecrets(localConfig, enableCache, enableFallback, fallbackPath, legacyFallbackPath, metadataPath, fallbackReadonly, fallbackOnly, exitOnWriteFailure, passphrase, nameTransformer, dynamicSecretsTTL)
+		fallbackOpts := fallbackOptions{
+			enable:             enableFallback,
+			path:               fallbackPath,
+			legacyPath:         legacyFallbackPath,
+			readonly:           fallbackReadonly,
+			exclusive:          fallbackOnly,
+			exitOnWriteFailure: exitOnWriteFailure,
+			passphrase:         passphrase,
+		}
+		// retrieve secrets
+		dopplerSecrets := fetchSecrets(localConfig, enableCache, fallbackOpts, metadataPath, nameTransformer, dynamicSecretsTTL)
+
+		mountPath := cmd.Flag("mount").Value.String()
+		mountFormat := cmd.Flag("mount-format").Value.String()
+		maxReads := utils.GetIntFlag(cmd, "mount-max-reads", 32)
+		// only auto-detect the format if it hasn't been explicitly specified
+		shouldAutoDetectFormat := !cmd.Flags().Changed("mount-format")
+		shouldMountFile := mountPath != ""
 
 		if preserveEnv {
-			utils.LogWarning("Ignoring Doppler secrets already defined in the environment due to --preserve-env flag")
+			if shouldMountFile {
+				utils.LogWarning("--preserve-env has no effect when used with --mount")
+			} else {
+				utils.LogWarning("Ignoring Doppler secrets already defined in the environment due to --preserve-env flag")
+			}
 		}
 
-		env := os.Environ()
+		originalEnv := os.Environ()
 		existingEnvKeys := map[string]bool{}
-		for _, envVar := range env {
+		for _, envVar := range originalEnv {
 			// key=value format
 			parts := strings.SplitN(envVar, "=", 2)
 			key := parts[0]
 			existingEnvKeys[key] = true
 		}
 
+		secrets := map[string]string{}
 		excludedKeys := []string{"PATH", "PS1", "HOME"}
-		for name, value := range secrets {
+		for name, value := range dopplerSecrets {
 			useSecret := true
-			for _, excludedKey := range excludedKeys {
-				if excludedKey == name {
-					useSecret = false
-					break
+			if !shouldMountFile {
+				// ignore secrets that might conflict with the environment
+				for _, excludedKey := range excludedKeys {
+					if excludedKey == name {
+						utils.LogDebug(fmt.Sprintf("Ignoring restricted secret %s", name))
+						useSecret = false
+						break
+					}
 				}
-			}
 
-			if useSecret && preserveEnv {
 				// skip secret if environment already contains variable w/ same name
-				if existingEnvKeys[name] == true {
+				if preserveEnv && existingEnvKeys[name] == true {
```

Comment:
Should this logic be flipped? If we're not mounting the file, we want to set `useSecret = false` when `preserveEnv == false` and there is an existing key matching.

Reply:
Hold on a minute, I've gone and confused myself. Please stand by

Reply:
Disregard, this logic is correct 👍